src/Win32_Interop/Win32_QFork.cpp

```diff
@@ -1170,9 +1170,14 @@ LPVOID AllocHeapBlock(size_t size, BOOL allocateHigh) {
     }
     int contiguousBlocksToAllocate = (int)(size / g_pQForkControl->heapBlockSize);
 
+    if (contiguousBlocksToAllocate > g_pQForkControl->availableBlocksInHeap) {
+        errno = ENOMEM;
+        return retPtr;
+    }
+
     size_t mapped = 0;
-    int startIndex = allocateHigh ? g_pQForkControl->availableBlocksInHeap - 1 : contiguousBlocksToAllocate - 1;
-    int endIndex = allocateHigh ? -1 : g_pQForkControl->availableBlocksInHeap - contiguousBlocksToAllocate + 1;
+    int startIndex = allocateHigh ? g_pQForkControl->availableBlocksInHeap - 1 : 0;
+    int endIndex = allocateHigh ? -1 : g_pQForkControl->availableBlocksInHeap;
```

Comment:
...but this one isn't. The problem is that the inner loop will test for up to contiguousBlocksToAllocate blocks starting from blockIndex, without checking for array boundaries. So endIndex must be set conservatively to avoid out-of-range access. So blockIndex should stop at g_pQForkControl->availableBlocksInHeap - contiguousBlocksToAllocate + 1  in the case of allocateHigh FALSE.

However the -1 bound for the high-to-low case is also broken.